pkg/puller/watcher.go

```diff
@@ -0,0 +1,184 @@
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				isCreate := event.Op&fsnotify.Create != 0
+				eventPath := filepath.Clean(event.Name)
+				isDataDir := filepath.Base(eventPath) == "..data"
+				// TODO: Should we use atomic integer or timestamp??
+				timeNow := time.Now()
+				if isDataDir && isCreate {
+					symlink, _ := filepath.EvalSymlinks(eventPath)
+					file, err := ioutil.ReadFile(filepath.Join(symlink, constants.ModelConfigFileName))
+					modelConfigs := make(modelconfig.ModelConfigs, 0)
+					if err != nil {
+						log.Println("Error in reading file", err)
+					}
+					err = json.Unmarshal([]byte(file), &modelConfigs)
+					if err != nil {
+						log.Println("unable to marshall for", event, "with error", err)
+					}
+					for _, modelConfig := range modelConfigs {
+						modelName := modelConfig.Name
+						modelSpec := modelConfig.Spec
+						log.Println("Name:", modelName, "Spec:", modelSpec)
+						oldModelInterface, ok := w.ModelTracker.Load(modelName)
+						if !ok {
+							w.ModelTracker.Store(modelName, ModelWrapper{
+								ModelSpec: &modelSpec,
+								Time:      timeNow,
+								Stale:     true,
+								Success:   false,
+							})
+						} else {
+							oldModel := oldModelInterface.(ModelWrapper)
+							isSame := false
+							if oldModel.ModelSpec != nil {
+								isSame = cmp.Equal(*oldModel.ModelSpec, modelSpec)
+								log.Println("same", isSame, *oldModel.ModelSpec, modelSpec)
+							}
+							if isSame {
+								// Need to store new time, maybe worth to have seperate map?
+								w.ModelTracker.Store(modelName, ModelWrapper{
+									ModelSpec: &modelSpec,
+									Time:      timeNow,
+									Stale:     false,
+									Success:   false,
+								})
+							} else {
+								w.ModelTracker.Store(modelName, ModelWrapper{
+									ModelSpec: &modelSpec,
+									Time:      timeNow,
+									Stale:     true,
+									Success:   false,
+								})
+							}
+						}
+					}
+					// TODO: Maybe make parallel and more efficient?
+					w.ModelTracker.Range(func(key interface{}, value interface{}) bool {
+						modelName, modelWrapper := key.(string), value.(ModelWrapper)
+						if modelWrapper.Time.Before(timeNow) {
+							w.ModelTracker.Delete(modelName)
+							channel, ok := p.ChannelMap[modelName]
+							if !ok {
+								log.Println("Model", modelName, "was never added to channel map")
+							} else {
+								event := EventWrapper{
+									ModelSpec:      nil,
+									LoadState:      ShouldUnload,
+									ShouldDownload: !modelWrapper.Success,
+								}
+								log.Println("Sending event", event)
+								channel.EventChannel <- event
+							}
+						} else {
+							if modelWrapper.Stale {
+								channel, ok := p.ChannelMap[modelName]
+								if !ok {
+									log.Println("Need to add model", modelName)
+									// TODO: Maybe have more workers per Channel?
+									channel = p.AddModel(modelName, w.numWorkers)
```

Comment:
I think we should only allow one worker per model channel, otherwise we are going to run into race condition when processing downloading and deleting events at same time.

Reply:
I set it for 1 right now, I'm not quite clear on the use-case for more than 1. but maybe some fancy go-routine `.Merge()` logic can be done?